DP-1232 One handler event on list.set()

aleksei-berezkin opened the pull request:
This is a member of a triple pull request. This one may look independent, but I didn't tested it separately from others, so please merge all of them or neither.

Dependent pull requests to be merged synchronously:
https://github.com/JetBrains/jetpad-projectional/pull/275
https://github.com/JetBrains/datapad/pull/36

glebleonov commented:
I see 11 usages of CollectionItemEvent.SET, but 40 usages of ADD and 32 usages of REMOVE. All these usages should be inspected and, most likely, refactored. @aleksei-berezkin, did you perform such work?

glebleonov commented on model/src/main/java/jetbrains/jetpad/model/collections/CollectionEventHandlers.java:
```diff
@@ -0,0 +1,33 @@
+
+import jetbrains.jetpad.model.event.EventHandler;
+
+public class CollectionEventHandlers {
```
I don't see any objection to place this class in test sources.

aleksei-berezkin replied:
Yes, I use them in other modules tests.

aleksei-berezkin commented:
I checked all usages of `CollectionItemEvent`, and the only place requiring changes was in Census (plz see dependent datapad request). Other places either regarded all three events fairly, or used collection adapters which is safe.

glebleonov commented on model/src/test/java/jetbrains/jetpad/model/collections/ObservableArrayListTest.java:
```diff
@@ -118,6 +119,15 @@ public void iteratorRemove() {
+  @Test
+  public void oneHandlerEventOnSet() {
+    addSampleItem();
+    CollectionEventHandlers.CountingHandler<String> counter = new CollectionEventHandlers.CountingHandler<>();
```
It is better to check for one set event explicitly.

aleksei-berezkin commented:
Checked usages of ADD, REMOVE and SET, found two places requiring changes, both of them in DataPad, updated the corresponding request.

Changes in last commit of this request: moved counting handler to test sources, added event kind check.

glebleonov commented:
Ok, prepare a squashed pull request, please.

aleksei-berezkin commented:
Please see one-commit request #84.
